src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -257,29 +350,170 @@ public void WriteValue(byte[] value)
         {
             if (value == null)
             {
-                this.writer.WriteNullValue();
+                this.utf8JsonWriter.WriteNullValue();
             }
             else
             {
-                this.writer.WriteBase64StringValue(value);
+                this.utf8JsonWriter.WriteBase64StringValue(value);
             }
             
             this.FlushIfBufferThresholdReached();
         }
 
         public void WriteRawValue(string rawValue)
+        {
+            this.WriteRawValueCore(rawValue);
+            this.FlushIfBufferThresholdReached();
+        }
+
+        /// <summary>
+        /// Implementation of WriteRawValue logic, used by both <see cref="WriteRawValue(string)"/>
+        /// and <see cref="WriteRawValueAsync(string)"/>.
+        /// </summary>
+        /// <param name="rawValue">The value to write.</param>
+        private void WriteRawValueCore(string rawValue)
         {
             if (rawValue == null)
             {
                 return;
             }
 
+            // We transcode and write the raw value directly to the buffer writer
+            // because Utf8JsonWriter.WriteRawValue is not available in .NET Core 3.1 and .NET Standard 2.0
+            // Writing the value manually means we also have to manually keep track of whether the separator
+            // should be written because Utf8JsonWriter is not aware of this write.
             // Consider using Utf8JsonWriter.WriteRawValue() in .NET 6+
             // see: https://github.com/OData/odata.net/issues/2420
 
-            this.Flush(); // ensure we don't write to the stream while there are still pending data in the buffer
-            this.writeStream.Write(Encoding.UTF8.GetBytes(rawValue));
-            this.FlushIfBufferThresholdReached();
+            // ensure we don't write to the buffer directly while there are still pending data in the Utf8JsonWriter buffer
+            this.CommitUtf8JsonWriterContentsToBuffer();
+            if (IsInArray() && !isWritingAtStartOfArray)
+            {
+                // Place a separator before the raw value if
+                // this is an array, unless this is the first item in the array.
+                this.bufferWriter.Write(itemSeparator.Slice(0, 1).Span);
+            }
+
+            this.bufferWriter.Write(Encoding.UTF8.GetBytes(rawValue));
+
+            // since we bypass the Utf8JsonWriter, we need to signal to other
+            // Write methods that a separator should be written first
+            if ((this.isWritingAtStartOfArray || this.isWritingConsecutiveRawValuesAtStartOfArray) || !this.IsInArray())
+            {
+                this.shouldWriteSeparator = true;
+            }
+
+            if (this.isWritingAtStartOfArray || this.isWritingConsecutiveRawValuesAtStartOfArray)

```

Comment:
```suggestion
            if (this.isWritingAtStartOfArray)
```
Otherwise we're setting `this.isWritingConsecutiveRawValuesAtStartOfArray` to `true` when it already has a value of `true`...